### ISSUE TITLE
cannot render .Rmarkdown file with `ggplotly`

### ISSUE THREAD

**Issue by lz100** [rstudio/blogdown#578]
Welcome to the blogdown GitHub repo 👋

To file a _bug report_, please follow these instructions carefully: <https://yihui.org/issue/#bug-reports>

## Checklist

When filing a _bug report_, please check the boxes below to confirm that you have provided us with the information we need. Have you:

- [x] [formatted your issue](https://yihui.org/issue/#please-format-your-issue-correctly) so it is easier for us to read?

- [x] used `blogdown::check_site()` function to verify your current project and followed all advices ?

- [x] included a minimal, self-contained, and reproducible example?

- [x] pasted the output from `xfun::session_info('blogdown')` in your issue?

- [x] upgraded all your packages to their latest versions (including your versions of R, the RStudio IDE, and relevant R packages)?

- [x] installed and tested your bug with the development version of the blogdown package using `remotes::install_github('rstudio/blogdown')`?

With develop version of blogdown, `.Rmarkdown` files cannot be transformed to `.markdown` files if I use `plotly::ggplotly(p)` inside. It stopped at `.knit.md` file and gives error:

```r
The extension tex_math_dollars is not supported for gfm
Error: pandoc document conversion failed with error 23
Execution halted
Error : Failed to render content/en/docs/test.Rmarkdown
```
The rmarkdown look like this

```
---
title: "Example Rmarkdown"
author: "xx"
date: "1/28/2021"
output: html_document
---

\```{r}
p <- ggplot2::ggplot()
p
plotly::ggplotly(p)
\```
```
(remove the \ to reproduce)

1. If I comment out `plotly::ggplotly(p)`, everything works and `.markdown` file is created. 

2. I try to change `.knit.md` to `.markdown`, the page is displayed but contains some raw HTML, seems post processing is 
needed and blogdown failed at that step.

3. I use {rmarkdown} to directly render to html, no problem.

```r
R version 4.0.3 (2020-10-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.5 LTS, RStudio 1.4.910

Locale:
  LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
  LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
  LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

Package version:
  base64enc_0.1.3   BH_1.75.0.0       blogdown_1.1.7    bookdown_0.21.6   digest_0.6.27     evaluate_0.14     glue_1.4.2       
  graphics_4.0.3    grDevices_4.0.3   highr_0.8         htmltools_0.5.1.1 httpuv_1.5.5      jsonlite_1.7.2    knitr_1.31       
  later_1.1.0.1     magrittr_2.0.1    markdown_1.1      methods_4.0.3     mime_0.9          promises_1.1.1    R6_2.5.0         
  Rcpp_1.0.6        rlang_0.4.10      rmarkdown_2.6     servr_0.21        stats_4.0.3       stringi_1.5.3     stringr_1.4.0    
  tinytex_0.29      tools_4.0.3       utils_4.0.3       xfun_0.20         yaml_2.2.1       

Hugo version: 0.79.0
```

**Comment by lz100** [comment]
problem solved by updating Rstudio which updated pandoc from 2.9 to 2.12.

**Comment by cderv** [comment]
Hi @lz100 , 

I don't think this is intended. I'll reopen so that we can check that and make the correct change. 

Thank you for the report

**Comment by yihui** [comment]
I didn't find a specific mention of when this extension was made available for the `gfm` output format, but my guess is Pandoc 2.10.1: https://github.com/jgm/pandoc/blob/master/changelog.md#pandoc-2101-2020-07-23 @cderv If you want to verify that, please feel free to. Thanks!

**Comment by cderv** [comment]
I did not fix it quickly the other day because I was under the impression it was trickier that it seems. So your fix and your question makes me look into that now.

So yes I checked and `tex_math_dollars` indeed appeared in Pandoc 2.10.1 when they switched the underlying markdown library for `gfm` and `commonmark`. Before that it was not supported.

But it is the same with `footnotes` so I think the fix is not enough. 
```powershell
> pandoc --version
pandoc.exe 2.9
Compiled with pandoc-types 1.20, texmath 0.12, skylighting 0.8.3
Default user data directory: C:\Users\chris\AppData\Roaming\pandoc
Copyright (C) 2006-2019 John MacFarlane
Web:  http://pandoc.org
This is free software; see the source for copying conditions.
There is no warranty, not even for merchantability or fitness
for a particular purpose
> pandoc -t gfm+footnotes
The extension footnotes is not supported for gfm
````

Also, know it errors like that only from Pandoc 2.8, but before it had just silently no effect.
For example with Pandoc 2.7.3
```
❯ pandoc -t gfm+footnotes
Text with a footnotes[^1]

[^1]: the foonote
^Z
Text with a footnotes\[1\]

1.  the foonote
```

Where we should get as with last Pandoc 2.11.4
````
❯ pandoc -t gfm+footnotes
Text with a footnotes[^1]

[^1]: the foonote
^Z
Text with a footnotes[^1]

[^1]: the foonote
````

To be exhaustive, here is the list of extensions supported before the change in 2.10.1 (in 2.9.2.1)
````
pandoc --list-extensions=gfm
+all_symbols_escapable
-ascii_identifiers
+auto_identifiers
+autolink_bare_uris
+backtick_code_blocks
-east_asian_line_breaks
+emoji
+fenced_code_blocks
+gfm_auto_identifiers
-hard_line_breaks
+intraword_underscores
+lists_without_preceding_blankline
+pipe_tables
+raw_html
-raw_tex
+shortcut_reference_links
-smart
+space_in_atx_header
+strikeout
+task_lists
````

And here is the list now (2.11.4) after some fixes in Pandoc since 2.10.1

````
pandoc --list-extensions=gfm
-ascii_identifiers
+autolink_bare_uris
-bracketed_spans
-definition_lists
-east_asian_line_breaks
+emoji
-fancy_lists
-fenced_code_attributes
-fenced_divs
-footnotes
+gfm_auto_identifiers
-hard_line_breaks
-implicit_figures
-implicit_header_references
+pipe_tables
-raw_attribute
+raw_html
-raw_tex
-smart
+strikeout
-subscript
-superscript
+task_lists
-tex_math_dollars
-attributes
-sourcepos
````


It means that before 2.10.1, Footnotes and TeX maths would not be correctly kept in the `.Rmarkdown` output. 

Can you remind me the aim of this conversion ? Why using `gfm` as a base format ? 
Should we try to output a "goldmark" markdown format from `markdown_strict` + extensions like [in the IDE](https://github.com/rstudio/rstudio/blob/ddcd7191ec89c4da00e77afae7e9f27e61e87c36/src/gwt/panmirror/src/editor/src/api/pandoc_format.ts#L397), also used in [hugodown ](https://github.com/r-lib/hugodown/blob/4ed6e098a51cc4707ab72ab04ed5874bf24d292c/R/md-document.R#L143)? 

Something like 

````
goldmark_format <- function() {
  paste(
    "markdown_strict",
    "all_symbols_escapable",
    "backtick_code_blocks",
    "fenced_code_blocks",
    "space_in_atx_header",
    "intraword_underscores",
    "lists_without_preceding_blankline",
    "shortcut_reference_links",
    "pipe_tables",
    "strikeout",
    "autolink_bare_uris",
    "task_lists",
    "definition_lists",
    "footnotes",
    "smart",
    "tex_math_dollars",
    "yaml_metadata_block",
    sep = "+"
  )
}
````

**Comment by yihui** [comment]
> Can you remind me the aim of this conversion ?

The aim is to convert Pandoc's Markdown to Hugo's, because the former is a superset of the latter, and we need to convert/remove Markdown features that are unique to Pandoc.

> Why using `gfm` as a base format ?

Because I didn't want to manage the task of figuring out which extensions should be enabled or disabled in the future. `gfm` is closer to Hugo's Markdown than `markdown_strict`, so I may have fewer extensions to maintain. For example, `fenced_divs` works with `gfm` but not `markdown_strict` (you can't enable this extension), and I feel the latter is too strict.

I don't have a perfect solution to this problem. I just added a global option `blogdown.markdown.output` so users can decide by themselves which format or extensions to use.

**Comment by cderv** [comment]
Thanks for the clarification. Adding an option is indeed a good solution to allow user's make there own choice. 

> For example, fenced_divs works with gfm but not markdown_strict (you can't enable this extension), and I feel the latter is too strict.

Not sure to follow the example. `fenced_divs` can be enabled with `gfm` since pandoc 2.10.1 only and  with`markdown_strict` since before. Also it is disabled by default for both. `markdown_strict` can accept almost all extensions (from what I understand).
However I see what you mean by 'figuring out which extensions should be enabled or disabled in the future'. 

And regarding the aim, this confirms what I thought. Thanks.

In summary using `gfm+tex_math_dollars+footnotes` will work only from Pandoc 2.10.1 because footnotes and tex_math_dollars are only available since this version. With previous Pandoc, it would output a different markdown document because Pandoc changed it. 

Anyway, your solution fixes in a way the issue because user get control. I'll look out for future issues on this and manage this pandoc part of the code.

**Comment by yihui** [comment]
I may have misunderstood something, but I want fenced Divs to be converted to `<div>`, e.g., `gfm` works fine:

```
pandoc -f markdown -t gfm
```

```
::: {.foo bar=2}
**b** c
:::
```

```html
<div class="foo" bar="2">

**b** c

</div>
```

However, I can't generate the `<div>` with `markdown_strict`:

```
pandoc -f markdown -t markdown_strict-fenced_divs
```

```
::: {.foo bar=2}
**b** c
:::
```

```
**b** c
```

```
pandoc -f markdown -t markdown_strict+fenced_divs
```

```
::: {.foo bar=2}
**b** c
:::
```

```
::: {.foo bar="2"}
**b** c
:::
```

I'm testing with Pandoc 2.11.4.

**Comment by cderv** [comment]
Oh I see. 

It is not `fenced_divs` who supports that for the output format. If you remove from `gfm`, you still get the result.
````
pandoc -f markdown -t gfm-fenced_divs
::: {.foo bar=2}
**b** c
:::
^Z
<div class="foo" bar="2">

**b** c

</div>
````

And also `markdown_strict` has `fenced_divs` disable by default all (so `-fenced_divs`)

I believe it is [`native_divs`](https://pandoc.org/MANUAL.html#extension-native_divs) which is enabled by default in `gfm` and can't be disabled
````
pandoc -f markdown -t markdown_strict+native_divs
::: {.foo bar=2}
**b** c
:::
^Z
<div class="foo" bar="2">

**b** c

</div>
````

**Comment by yihui** [comment]
Okay. I see now. Thanks for the explanation!
